subprojects/scala/src/integTest/groovy/org/gradle/scala/compile/CachedScalaCompileIntegrationTest.groovy

```diff
@@ -0,0 +1,189 @@
+        withBuildCache().succeeds ':compileJava', compilationTask
+
+        then:
+        compiledJavaClass.exists()
+        compiledScalaClass.exists()
+
+        when:
+        withBuildCache().succeeds ':clean', ':compileJava'
+
+        then:
+        skipped ':compileJava'
+
+        when:
+        // This line is crucial to expose the bug
+        // When doing this and then loading the classes for
+        // compileScala from the cache the compiled java
+        // classes are replaced and recorded as changed
+        compiledJavaClass.makeOlder()
+        withBuildCache().succeeds compilationTask
+
+        then:
+        skipped compilationTask
+
+        when:
+        file('src/main/java/RequiredByScala.java').text = """
+            public class RequiredByScala {
+                public static void printSomethingNew() {
+                    java.lang.System.out.println("Hello from Java");
+                    // Different
+                }
+            }
+        """
+        file('src/main/scala/UsesJava.scala').text = """
+            class UsesJava {
+                def printSomething(): Unit = {
+                    RequiredByScala.printSomethingNew()
+                    // Some comment
+                }
+            }
+        """
+
+        withBuildCache().succeeds compilationTask
+
+        then:
+        compiledJavaClass.exists()
+        compiledScalaClass.exists()
+    }
+
+    def "incremental compilation works with caching"() {
+        def warmupDir = testDirectory.file('warmupCache')
+        setupProjectInDirectory(warmupDir)
+        warmupDir.file('settings.gradle') << localCacheConfiguration()
+
+        def classes = new ScalaCompilationFixture(warmupDir)
+        classes.baseline()
+        classes.classDependingOnBasicClassSource.change()
+
+        when:
+        executer.inDirectory(warmupDir)
+        withBuildCache().succeeds compilationTask
+
+        then:
+        classes.all*.compiledClass*.exists().every()
+
+        when:
+        warmupDir.deleteDir()
+        setupProjectInDirectory(testDirectory)
+        classes = new ScalaCompilationFixture(testDirectory)
+        classes.baseline()
+        withBuildCache().succeeds compilationTask
+
+        then:
+        executedAndNotSkipped compilationTask
+
+        when:
+        classes.classDependingOnBasicClassSource.change()
+        withBuildCache().succeeds compilationTask
+
+        then:
+        skipped compilationTask
+
+        when:
+        classes.independentClassSource.change()
+        withBuildCache().succeeds compilationTask
+
+        then:
+        executedAndNotSkipped compilationTask
+        assertAllRecompiled(classes.allClassesLastModified, old(classes.allClassesLastModified))
+    }
+
+    private static void assertAllRecompiled(List<Long> lastModified, List<Long> oldLastModified) {
+        [lastModified, oldLastModified].transpose().each { after, before ->
+            assert after != before
+        }
+    }
+
+}
```

Comment:
`incremental compilation works with caching` is more interesting for Java, we do not have an incremental Groovy compiler AFAIK. I wouldn't bother to unify Joint compilation test right now. In my opinion the amount of shared code does not justify the investment.